007-OFEP-flag-change-events.md

```diff
@@ -0,0 +1,74 @@
+
+Examples:
+
+- A web client's banner color could be updated when an associated flag is changed, without any user action or a page reload.
+- A service could subscribe to a flag that controls its global log-level.
+- A subsystem could be restarted with the new flag value.
+
+## Design
+
+The provider interface and the OpenFeature client would be extended to have new functionality to register handlers for a set of events defined by the SDK (`ProviderEvents`). When a _application author_ registers a handler on a client, the client the client maintains this handler. The provider emits events or runs a callback indicating that it received a certain event, optionally providing data associated with that event. The callbacks registered with the client are then invoked with this data (`EventData`).
+
+In the case of the aforementioned flag systems, flags are not actually evaluated when configurations are changed. One reason is that no dynamic context can be reasonably provided in the case of events, since the event is driven by a change in the flag management system, not a user-action. In these cases, the `EventData` would not contain flag values and the application author would have to evaluate flags in the registered handler. This is consistent with the event APIs already existing in those systems.
+
+How it looks implemented in a Provider:
+
+```ts
+import { Provider, EventingProvider, ProviderEvents } from './types';
+
+/**
+ * The No-op provider is set by default, and simply always returns the default value.
+ */
+class MyEventingProvider implements Provider, EventingProvider {
+  readonly metadata = {
+    name: 'My Eventing Provider',
+  } as const;
+
+  // ...
+
+  // the SDK listens for events, and fires associated handlers the application-author adds.
+  readonly events = new EventEmitter();
+
+  // pollDataSource a conceptual method specific to this provider that fires a callback if the flag source-of-truth of this provider changes.
+  this.pollDataSource((newFlagData) => this.events.emit(ProviderEvents.ConfigurationChanged, newFlagData))
+
+  // ...
+
+```
+
+How it looks for an `application author`:
+
+```ts
+OpenFeature.setProvider(new MyEventingProvider());
+
+const client = OpenFeature.getClient();
+
+// subscribe to ProviderEvents.ConfigurationChanged events
+client.addHandler(ProviderEvents.ConfigurationChanged, (eventData: EventData | undefined) => {
```

Comment:
I think I'd like the ability to scope down this handler to only events of a particular key.
```java
logger.setLogLevel(client.getInteger('log_level', 4));
client.onChange('log_level', (i) -> logger.setLogLevel(i));
```

Reply:
I understand that, and it many cases it should be possible. It's not possible in all cases however, since there are SDKs which only send events indicating a general change occurred, not the flags changed.

In such cases, it could mean "run all handlers" or "run no handlers" or the provider might provide a configuration option in its constructor to choose between these behaviors.

Any takes on this?

Reply:
> I understand that, and it many cases it should be possible. It's not possible in all cases however, since there are SDKs which only send events indicating a general change occurred, not the flags changed.

Isn't that where provider features could come into play?

Reply:
> Isn't that where provider features could come into play?

Provider features don't address the above questions totally. They will give the SDK an idea what this provider supports, but we still have the same fundamental question. Do we fire the handler or not? Do we make it configurable? The handler still has to do or not do _something_.

Reply:
My 2 cents is that we should make implementation of an API like @justinabrahms describes a userland concern, built on top of this core capability, rather than something that's supported directly by the spec/SDKs